doc/readbits.8
.TH READBITS "8" "January 2016" "sama5d2_utils\-0.92" SAMA5D3_UTILS
.SH NAME
readbits \- Read GPIO bit value using sysfs
.SH SYNOPSIS
.B readbits
[\fI\-b BN\fR] [\fI\-h\fR] [\fI\-i\fR] [\fI\-p PORT\fR] [\fI\-r\fR]
[\fI\-u\fR] [\fI\-U\fR] [\fI\-v\fR] [\fI\-V\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Reads a GPIO line (single bit) specified by \fIBN\fR via the Linux
kernel's gpiolib interface in sysfs.
.PP
The Atmel AT91 family (including the SAMA5 family) name each GPIO line as
follows: P<letter><num> where <letter> starts with 'A' and continues
to 'C', or 'E' depending on the SoC. And <num> is a number starting
at 0 up to 31, with the maximum value sometimes less than 31 depending on
the SoC. So examples include: PA0, PB31, PD5 and PD31.
.PP
GPIO lines can have many controls and be multiplexed with input and output
lines of other macrocells in a SoC. The Linux kernel provides the gpiolib
subsystem to hide individual SoC GPIO complexity and provide general
functionality. A relatively simple sysfs interface is provided by gpiolib
under the /sys/class/gpio directory. The interface is described
in <kernel_src>/Documentation/gpio.txt in the section titled: "Sysfs
Interface for Userspace (OPTIONAL)".
.PP
A GPIO line can be specified in one of three ways to this utility:
.TP
\fI\-b BN\fR only
Use the manufacturer's name (Atmel in this case) for the \fI\-b BN\fR option.
The letters can be in upper or lower case and the leading 'P' can be omitted.
Examples: '\-b PA5', '\-b pa5', '\-b a5' and '\-b PC31' .
.TP
\fI\-b BN\fR plus \fI\-p PORT\fR
In this form the bank letter is given to \fI\-p PORT\fR (e.g. '\-p B') and
the GPIO line number within that bank is given to
\fI\-b BN\fR (e.g. '\-b 5') . Combining these examples PB5 is the specified
GPIO line.
.TP
\fI\-p PORT\fR where \fIPORT\fR is a number
In this form the number given to \fI\-p PORT\fR (e.g. '\-p 35')
is a kernel GPIO number. Kernel GPIO numbers can be considered as a Linux
kernel experiment that failed in the case of Atmel SoCs. Worse, the
mapping between kernel GPIO numbers and Atmel GPIO naming has changed.
Best to avoid this form.
.PP
Since this utility uses a generic interface within the Linux kernel, it
should work with any processors that have GPIOs. The user's challenge is
to work out how that processor's GPIO naming conventions map to the
Linux kernel's GPIO numbers.
.SH OPTIONS
.TP
\fB\-b\fR \fIBN\fR
where \fIBN\fR either starts with a letter and is the manufacturer's GPIO
name (Atmel in this case) or it starts with a number. In the letter case
a leading 'P' is assumed if not given. The letter can be in upper or lower
case and must be followed by a number; examples: '\-b PA3', '\-b a3'
and '\-b PC31'. In the case where a number is given first (e.g. '\-b 3')
then that number is assumed to be the GPIO line within the bank
specified by the accompanying \fI\-p PORT\fR option.
.TP
\fB\-h\fR
print out usage message then exit.
.TP
\fB\-i\fR
ignore setting the line direction to input prior to reading the GPIO line
value. The default action is to set the line direction to input prior to
reading the GPIO line value. If that GPIO line is known to already be set
for input (e.g. due to a successful, prior call to readbits for the same
GPIO line) then a little time can be saved by not setting the line
direction again.
.TP
\fB\-p\fR \fIPORT\fR
if \fIPORT\fR  is a letter then it is assumed to be a PIO bank. For example
if \fIPORT\fR is 'B' or 'b' then that is Atmel PIO bank 'PB'. In this case
the \fI\-b BN\fR option is also expected where \fIBN\fR is a number, for
example '\-b 3'. Together these examples would specify PIO line PB3.
If \fIPORT\fR is a number then it is interpreted as a Linux kernel GPIO
number. Prior to the lk 3.4 series PA0 mapped to the kernel GPIO number of
32; thereafter is mapped to 0 . If practise, kernel GPIO numbers should
be avoided.
.TP
\fB\-r\fR
if successful this utility either prints out '0' or '1' reflecting the
logical state of the GPIO line. If \fI\-r\fR option is given once (and
this is also the default action) the exit status is 0 in both cases. The
exit status will typically be 1 if the utility is unsuccessful. If \fI\-r\fR
option is used twice then the exit status also reflects the logical
state of the GPIO line (i.e. either 0 (for low) or 1 (for high)).
.TP
\fB\-u\fR
this option causes the specified GPIO line to be "unexported" prior to other
operations by this utility. This may be required if the specified GPIO line
was left in the "exported" state (e.g. a previous incantation of this utility
used the \fI\-U\fR option). If this operation fails then an error message is
output and the utility continues. See the kernel documentation file (gpio.txt)
for more information about exported and unexported states. At boot up
typically all available GPIO lines are in the "unexported" state.
.TP
\fB\-U\fR
this option causes the specified GPIO line to be left "exported" when this
utility exits. The default action is to "unexport" the specified GPIO line
prior to this utility exiting.
.TP
\fB\-v\fR
increase the level of verbosity, (i.e. debug output). Additional output
caused by this option is sent to stderr.
.TP
\fB\-V\fR
print the version string and then exit.
.SH NOTES
Atmel generic GPIO lines may be multiplexed with other functions. Examples
of these other functions include GPIO lines used for serial ports (e.g. TXD,
RXD and RTS), I2C, SPI, 100_Mbps_ethernet and interrupt lines. The readbits
utility should be used with care since if the given GPIO line is "in use"
then after an invocation of readbits is finished then that GPIO line will be
left in generic mode as an input. That may cause the previous function that
the GPIO line was being used for to fail. Some versions of the Linux kernel
prohibit utilities like readbits touching reserved GPIO lines. However in
recent kernels (which have recently introduced "Device Tree" (open firmware)
configuration) everything seems to be allowed. That may (and should) change
to impose restrictions on other utilities interfering with reserved GPIO
lines.
.SH EXAMPLES
First check if some GPIO lines are in use:
.PP
   a5d2_pio_status \-S
.br
       PIN   PIO_A        PIO_B        PIO_C        PIO_D        PIO_E
.br
       0:    GPIO         GTX0         ETX0         MCI0_CDA     A0/NBS0
.br
       .....
.br
       31:   GPIO         DTXD         GPIO         GPIO         GPIO
.PP
This indicates the PB31 is being used for the debug serial port TXD line
and the other GPIO lines are "free". It is possible that some of those other
GPIO lines are in use, for example driving small LEDs or reading the
state of a switch. So more checking may be warranted. Let us assume PA31
is free:
.PP
   readbits \-b pa31
.br
     1
.PP
Note that since the Atmel boot sequence sets most GPIO lines to have an
internal pull-up resistor, most disconnected GPIO lines will yield 1 when
read. To help script writers, the value output to stdout (i.e. '1' or '0')
can also be set as the exit status value by using the \fI\-r\fR option:
.PP
   readbits \-b PA31 \-r
.br
     1
.PP
   echo $?
.br
     1
.PP
.SH EXIT STATUS
The exit status of readbits is 0 when it is successful except if the
\fI\-r\fR option is given. If readbits is successful and the \fI\-r\fR option
has been given then the exit status is 0 or 1 reflecting the GPIO line
state that has been read. If readbits fails then the exit status is most
likely to be 1.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2016 Douglas Gilbert
.br
This software is distributed under a FreeBSD license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B setbits(sama5d2_utils)


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. See README file.

Changelog for sama5d2_utils-0.90 [20160106] [svn: r2]
  - initial



src/readbits.c
/*
 * Copyright (c) 2010-2013 Douglas Gilbert.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

/*****************************************************************
 * readbits.c
 *
 * Utility for reading a GPIO line on a AT91SAM9G20/25/45 and SAMA5D3 in
 * Linux. Uses sysfs interface. The target hardware is a FoxG20, the Aria
 * G25 and SAMA5D3 family. This utility mimics the actions of a utility of
 * the same name for the FoxLX board. The Aria G25, FoxG20 and FoxLX
 * boards are made by Acme Systems. The SoCs are made by Atmel.
 *
 ****************************************************/

#define _XOPEN_SOURCE 500

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>
#include <errno.h>


static const char * version_str = "1.07 20131124";

#define EXPORT_FILE "/sys/class/gpio/export"
#define UNEXPORT_FILE "/sys/class/gpio/unexport"
#define GPIO_BASE_FILE "/sys/class/gpio/gpio"
#define PIO_BASE_FILE "/sys/class/gpio/pio"
#define GPIO_BANK_ORIGIN "/sys/class/gpio/gpiochip0"
/* Earlier kernels (G20+G25) offset GPIO numbers by 32 so
 * /sys/class/gpio/gpiochip32 (a directory) was the lowest numbered bank
 * and corresponded to PIOA (therefore PA0 was gpio number 32). By lk 3.7
 * /sys/class/gpio/gpiochip0 existed and corresponded to PIOA (therefore
 * PA0 becomes gpio number 0).
 * In lk 3.8 the sysfs gpio file names changed from /sys/class/gpio/gpio37
 * to /sys/class/gpio/pioB5 however the export and unexport files still
 * expect GPIO numbers.
 */


static int verbose = 0;
static const char * gpio_name = NULL;


static void
usage(void)
{
    fprintf(stderr, "Usage: "
            "readbits [-b BN] [-h] [-i] [-p PORT] [-r] [-u] [-U] [-v] [-V]\n"
            "  where:\n"
            "    -b BN        bit number within a port (0 to 31). Also\n"
            "                 accepts prefix like 'pb' or just 'b' for "
            "PORT.\n"
            "    -h           print usage message\n"
            "    -i           ignore line direction before reading (def: "
            "make input)\n"
            "    -p PORT      port ('a' to 'e') or gpio kernel line number "
            "(0 or more)\n"
            "    -r           print bit value to stdout (which is default "
            "action)\n"
            "                 used twice: exit value 0 for low, 1 for "
            "high\n"
            "    -u           unexport gpio line prior to reading bit\n"
            "    -U           leave line exported on exit\n"
            "    -v           increase verbosity (multiple times for more)\n"
            "    -V           print version string then exit\n\n"
            "Read GPIO line state on AT91SAM9G20+G25/SAMA5D3 using sysfs. "
            "For the Aria G25\nand FoxG20 boards. This utility is similar "
            "to a FoxLX utility of the same\nname. Note: in earlier Linux "
            "kernels gpio kernel line numbers started at\n32 (for PA0), "
            "recent kernels start at 0 (for PA0), use '-vv' to check.\n"
            "Example: 'readbits -b PC7'\n");
}

static char *
get_best_gpio_name(int knum, char * b, int blen)
{
    if (gpio_name)
        snprintf(b, blen, "%s [kn=%d]", gpio_name, blen);
    else
        snprintf(b, blen, "knum=%d", knum);
    return b;
}


int
main(int argc, char ** argv)
{
    struct stat sb;
    int opt, k, exp_fd, unexp_fd, direction_fd, val_fd;
    int bn = -1;
    int ignore_dir = 0;
    int origin0 = 0;
    int read_val = 0;
    int knum = -1;
    int unexport = 0;
    int exported_on_exit = 0;
    int exported = 0;
    int ret = -1;
    const char * cp;
    char b[256];
    char base_dir[128];
    char ch;
    char bank = '\0';

    while ((opt = getopt(argc, argv, "b:hip:ruUvV")) != -1) {
        switch (opt) {
        case 'b':
            cp = optarg;
            if (isalpha(cp[0])) {
                if ('P' == toupper(*cp))
                    ++cp;
                ch = toupper(*cp);
                if ((ch >= 'A') && (ch <= 'E'))
                    bank = ch;
                else {
                    fprintf(stderr, "'-b' expects a letter ('A' to 'E')\n");
                    exit(EXIT_FAILURE);
                }
                gpio_name = optarg;
                ++cp;
            }
            k = atoi(cp);
            if ((k < 0) || (k > 31)) {
                fprintf(stderr, "'-b' expects a bit number from 0 to 31\n");
                exit(EXIT_FAILURE);
            }
            bn = k;
            break;
        case 'h':
            usage();
            exit(EXIT_SUCCESS);
            break;
        case 'i':
            ++ignore_dir;
            break;
        case 'p':
            if (isalpha(*optarg)) {
                ch = toupper(*optarg);
                if ((ch >= 'A') && (ch <= 'E'))
                    bank = ch;
                else {
                    fprintf(stderr, "'-p' expects a letter ('A' to 'E') or "
                            "a number\n");
                    exit(EXIT_FAILURE);
                }
            } else if (isdigit(*optarg)) {
                k = atoi(optarg);
                if ((k < 0) || (k > 511)) {
                    fprintf(stderr, "'-p' expects a letter or a number "
                            "0 or greater\n");
                    exit(EXIT_FAILURE);
                }
                knum = k;
            } else {
                fprintf(stderr, "'-p' expects a letter ('A' to 'E') or "
                        "a number\n");
                exit(EXIT_FAILURE);
            }
            break;
        case 'u':
            ++unexport;
            break;
        case 'U':
            ++exported_on_exit;
            break;
        case 'r':
            ++read_val;
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            printf("%s\n", version_str);
            exit(EXIT_SUCCESS);
        default: /* '?' */
            usage();
            exit(EXIT_FAILURE);
        }
    }
    if (optind < argc) {
        if (optind < argc) {
            for (; optind < argc; ++optind)
                fprintf(stderr, "Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            exit(EXIT_FAILURE);
        }
    }

    if (! ((knum >= 0) || ((bank >= 'A') && (bn >= 0)))) {
        fprintf(stderr, "Expect either '-p PORT' or '-b BN'\n");
        usage();
        exit(EXIT_FAILURE);
    }
    if (stat(GPIO_BANK_ORIGIN, &sb) >= 0) {
        if (verbose > 1)
            fprintf(stderr, "%s found so kernel pin numbers start at 0 "
                    "(for PA0)\n", GPIO_BANK_ORIGIN);
        origin0 = 1;
    } else if (verbose > 2)
        fprintf(stderr, "%s not found so kernel pin numbers start at 32 "
                "(for PA0)\n", GPIO_BANK_ORIGIN);
    if (knum < 0) {
        knum = ((! origin0) + bank - 'A') * 32 + bn;
        if (verbose)
            fprintf(stderr, "%c%d becomes kernel pin number %d\n", bank,
                    bn, knum);
    } else if ((! origin0) && (knum < 32)) {
        fprintf(stderr, "since %s not found assume kernel pin numbers "
                "start at 32\n(for PA0) so %d is too low\n",
                GPIO_BANK_ORIGIN, knum);
        exit(EXIT_FAILURE);
    }
    if ('\0' == bank) {
        bank = (char)((knum / 32) + 'A' - (! origin0));
        bn = knum % 32;
    }

    exp_fd = -1;
    unexp_fd = -1;
    direction_fd = -1;
    val_fd = -1;
    exp_fd = open(EXPORT_FILE, O_WRONLY);
    if (exp_fd < 0) {
        perror(EXPORT_FILE);
        goto bad;
    }
    unexp_fd = open(UNEXPORT_FILE, O_WRONLY);
    if (unexp_fd < 0) {
        perror(UNEXPORT_FILE);
        goto bad;
    }

    if (unexport) {
        snprintf(b, sizeof(b), "%d", knum);
        if (pwrite(unexp_fd, b, strlen(b), 0) < 0) {
            fprintf(stderr, "Unable to unexport %s: %s\n",
                    get_best_gpio_name(knum, b, sizeof(b)), strerror(errno));
            fprintf(stderr, "continue ...\n");
        }
    }
    snprintf(b, sizeof(b), "%d", knum);
    if (pwrite(exp_fd, b, strlen(b), 0) < 0) {
        fprintf(stderr, "Unable to export %s (already in use?): %s\n",
                get_best_gpio_name(knum, b, sizeof(b)), strerror(errno));
        goto bad;
    }
    exported = 1;
    /* check if /sys/class/gpio/gpio<knum> directory exists */
    snprintf(base_dir, sizeof(base_dir), "%s%d", GPIO_BASE_FILE, knum);
    if (stat(base_dir, &sb) >= 0) {
        if (verbose > 1)
            fprintf(stderr, "%s found so continue in original manner\n",
                    base_dir);
    } else {
        if (verbose > 2)
            fprintf(stderr, "%s not found, now check for pinctrl "
                    "convention\n", base_dir);
        /* check if /sys/class/gpio/pio<bank><bn> directory exists */
        snprintf(base_dir, sizeof(base_dir), "%s%c%d", PIO_BASE_FILE, bank,
                 bn);
        if (stat(base_dir, &sb) >= 0) {
            if (verbose > 1)
                fprintf(stderr, "%s found so pinctrl convention\n", base_dir);
        } else {
            fprintf(stderr, "Unable to find sysfs directory %s (for "
                    "direction)\n", base_dir);
            goto bad;
        }
    }

    snprintf(b, sizeof(b), "%s/direction", base_dir);
    direction_fd = open(b, O_WRONLY);
    if (direction_fd < 0) {
        fprintf(stderr, "Open %s: %s\n", b, strerror(errno));
        goto bad;
    }
    if (0 == ignore_dir) {
        if (pwrite(direction_fd, "in", 2, 0) < 0) {
            fprintf(stderr, "Unable to write 'in' to direction_fd: %s\n",
                    strerror(errno));
            goto bad;
        }
    }
    snprintf(b, sizeof(b), "%s/value", base_dir);
    val_fd = open(b, O_RDONLY);
    if (val_fd < 0) {
        fprintf(stderr, "Open %s: %s\n", b, strerror(errno));
        goto bad;
    }
    k = pread(val_fd, b, 1, 0);
    if (k < 0) {
        fprintf(stderr, "pread() failed: %s\n", strerror(errno));
        goto bad;
    } else if (0 == k) {
        fprintf(stderr, "pread could not find a value\n");
        goto bad;
    }
    printf("%c\n", b[0]);
    if (read_val > 1)
        ret = ('0' !=  b[0]);
    else
        ret = 0;

bad:
    if (val_fd >= 0)
        close(val_fd);
    if (direction_fd >= 0)
        close(direction_fd);
    if (unexp_fd >= 0) {
        if (exported && (! exported_on_exit)) {
            snprintf(b, sizeof(b), "%d", knum);
            k = pwrite(unexp_fd, b, strlen(b), 0);
            /* assign to k to suppress warning */
        }
        close(unexp_fd);
    }
    if (exp_fd >= 0)
        close(exp_fd);
    if (ret)
        return 1;
    else
        return 0;
}
